### PR TITLE
Support  having clause in new plan

### PR DESCRIPTION
**Pull request by shenli**
@coocood @qiuyesuifeng

### CODE REVIEW

File: optimizer/plan/planbuilder.go [hunk 1]
```diff
@@ -121,9 +126,22 @@ func (b *planBuilder) extractSelectAgg(sel *ast.SelectStmt) []*ast.AggregateFunc
...
 			item.Expr = n.(ast.ExprNode)
+			// If item is PositionExpr, we need to rebind it.
+			// For PositionExpr will refer to a ResultField in fieldlist.
+			// After extract AggExpr from fieldlist, it may be changed (See the code above).
+			if pe, ok := item.Expr.(*ast.PositionExpr); ok {
+				pe.Refer = sel.GetResultFields()[pe.N-1]
+			}
+		}
+	}
+	if sel.Having != nil {
```
Comment (coocood) [inline optimizer/plan/planbuilder.go]: `having` is the process before `order by`, so I think it's better to extract from `having` before `order by`.

File: optimizer/plan/planbuilder.go [hunk 2]
```diff
@@ -77,6 +77,11 @@ func (b *planBuilder) detectSelectAgg(sel *ast.SelectStmt) bool {
 			return true
 		}
 	}
+	if sel.Having != nil {
+		if ast.HasAggFlag(sel.Having.Expr) {
+			return true
+		}
+	}
```
Comment (coocood) [inline optimizer/plan/planbuilder.go]: Should also detect `order by`?

File: optimizer/resolver.go [hunk 3]
```diff
@@ -429,7 +467,12 @@ func (nr *nameResolver) resolveColumnInResultFields(cn *ast.ColumnNameExpr, rfs
 	var matched *ast.ResultField
 	for _, rf := range rfs {
 		matchAsName := cn.Name.Name.L == rf.ColumnAsName.L
-		matchColumnName := rf.ColumnAsName.L == "" && cn.Name.Name.L == rf.Column.Name.L
+		var matchColumnName bool
+		if ctx.checkOriginName {
+			matchColumnName = cn.Name.Name.L == rf.Column.Name.L
+		} else {
+			matchColumnName = rf.ColumnAsName.L == "" && cn.Name.Name.L == rf.Column.Name.L
```
Comment (coocood) [inline optimizer/resolver.go]: we can just remove `rf.ColumnAsName.L == ""`, so we don't need  `checkOriginName`.

File: optimizer/resolver.go [hunk 4]
```diff
@@ -159,9 +168,21 @@ func (nr *nameResolver) Enter(inNode ast.Node) (outNode ast.Node, skipChildren b
 	return inNode, false
 }
 
+/*
+	defer func() {
+		if cn.Refer != nil {
+			ctx.groupBy = append(ctx.groupBy, cn.Refer)
+		}
+	}()
+*/
```
Comment (coocood) [inline optimizer/resolver.go]: What does this commented code means?

File: optimizer/resolver.go [hunk 5]
```diff
@@ -78,10 +78,14 @@ type resolverContext struct {
...
 	// When visiting having, only fieldList and groupBy fields are available.
 	inHaving bool
+	// When visiting having, checks if the expr is a aggregate functions expr.
+	inHavingAgg bool
 	// OrderBy clause has different resolving rule than group by.
 	inOrderBy bool
 	// When visiting column name in ByItem, we should know if the column name is in an expression.
 	inByItemExpression bool
+	// When resolve having ident from fieldlist and can not find it, resolve the second time and ignore asname when compare name.
+	checkOriginName bool
```
Comment (coocood) [inline optimizer/resolver.go]: This can be removed.

File: optimizer/plan/planbuilder.go [hunk 6]
```diff
@@ -156,6 +177,9 @@ func (b *planBuilder) buildSelect(sel *ast.SelectStmt) Plan {
 		if hasAgg {
 			p = b.buildAggregate(p, aggFuncs, nil)
 		}
+		if sel.Having != nil {
+			p = b.buildHaving(p, sel.Having)
+		}
```
Comment (coocood) [inline optimizer/plan/planbuilder.go]: Having should be build after select fields.

File: optimizer/optimizer.go [hunk 7]
```diff
@@ -66,9 +66,9 @@ type supportChecker struct {
 
 func (c *supportChecker) Enter(in ast.Node) (ast.Node, bool) {
 	switch ti := in.(type) {
-	case *ast.SubqueryExpr, *ast.HavingClause:
+	case *ast.SubqueryExpr:
 		c.unsupported = true
-	case *ast.GroupByClause:
+	case *ast.GroupByClause, *ast.HavingClause:
```
Comment (coocood) [inline optimizer/optimizer.go]: We don't need to list supported type.

File: optimizer/resolver.go [hunk 8]
```diff
@@ -78,10 +78,14 @@ type resolverContext struct {
 	inGroupBy bool
 	// When visiting having, only fieldList and groupBy fields are available.
 	inHaving bool
+	// When visiting having, checks if the expr is a aggregate functions expr.
```
Comment (qiuyesuifeng) [inline optimizer/resolver.go]: an aggregate

File: optimizer/resolver.go [hunk 9]
```diff
@@ -333,13 +361,23 @@ func (nr *nameResolver) resolveColumnNameInContext(ctx *resolverContext, cn *ast
 	}
 	if ctx.inHaving {
 		// First group by, then field list.
-		if nr.resolveColumnInResultFields(cn, ctx.groupBy) {
+		if nr.resolveColumnInResultFields(ctx, cn, ctx.groupBy) {
 			return true
 		}
-		return nr.resolveColumnInResultFields(cn, ctx.fieldList)
+		found := nr.resolveColumnInResultFields(ctx, cn, ctx.fieldList)
+		if nr.Err != nil || found {
```
Comment (qiuyesuifeng) [inline optimizer/resolver.go]: We should distinct check `nr.Err ! = nil` and `found`.

File: optimizer/resolver.go [hunk 10]
```diff
@@ -333,13 +361,23 @@ func (nr *nameResolver) resolveColumnNameInContext(ctx *resolverContext, cn *ast
...
+			return found
+		}
+		ctx.checkOriginName = true
+		found = nr.resolveColumnInResultFields(ctx, cn, ctx.fieldList)
+		ctx.checkOriginName = false
+		if found || !ctx.inHavingAgg || nr.Err != nil {
```
Comment (qiuyesuifeng) [inline optimizer/resolver.go]: Check err first, then the value seems better.

**Comment by shenli** [comment]
PTAL @coocood @qiuyesuifeng

**Comment by coocood** [comment]
LGTM

**Comment by shenli** [comment]
@qiuyesuifeng PTAL

**Comment by qiuyesuifeng** [comment]
LGTM

**Pull request merged**